ledger.go

```diff
@@ -937,42 +936,64 @@ func (l *Ledger) query() {
 		votes = append(votes, &response.vote)
 	}
 
+ValidateVotes:
 	for _, vote := range votes {
-		// Filter nil blocks
+		vote := vote.(*finalizationVote)
+
+		// Filter nil block proposals.
 		if vote.block == nil {
-			continue
+			continue ValidateVotes
 		}
 
-		// Filter blocks with unexpected block height
+		// Filter block proposals at an unexpected height.
 		if vote.block.Index != current.Index+1 {
 			vote.block = nil
-			continue
+			continue ValidateVotes
 		}
 
-		// Filter blocks with at least 1 missing tx
+		// Filter block proposals containing transactions which our node has not
+		// locally archived, and mark them as missing.
 		if l.transactions.BatchMarkMissing(vote.block.Transactions...) {
 			vote.block = nil
-			continue
+			continue ValidateVotes
+		}
+
+		transactions, err := l.transactions.BatchFind(vote.block.Transactions)
+		if err != nil {
+			vote.block = nil
+			continue ValidateVotes
 		}
 
+		// Validate the height recorded on transactions inside the block proposal.
+		for _, tx := range transactions {
```

Comment:
We should add a test that checks if this works (e.g. modify one of the node's latest block height, and then check to make sure the node's proposed block never gets accepted):

![2019-12-06_11-01](https://user-images.githubusercontent.com/1431045/70292014-cb290900-1817-11ea-88b4-d7d063ca3a3f.png)

Reply:
Gotcha, will work on it.